lib/vauth/digest.c

```diff
@@ -709,7 +709,7 @@ static CURLcode auth_create_digest_http_message(
     digest->nc = 1;
 
   if(!digest->cnonce) {
-    char cnoncebuf[33];
+    char cnoncebuf[17];
     result = Curl_rand_hex(data, (unsigned char *)cnoncebuf,
```

Comment:
This code first generates a hex string, then l.718 encodes this hex string as base64. Why not simply encode 12 bytes of binary data from Curl_rand_bytes and then base64 encode it, resulting in exactly 16 bytes of cnonce with 96bits of randomness?

Reply:
Excellent point. Something like this:

~~~diff
--- a/lib/vauth/digest.c
+++ b/lib/vauth/digest.c
@@ -707,13 +707,17 @@ static CURLcode auth_create_digest_http_message(
   memset(hashbuf, 0, sizeof(hashbuf));
   if(!digest->nc)
     digest->nc = 1;
 
   if(!digest->cnonce) {
-    char cnoncebuf[33];
-    result = Curl_rand_hex(data, (unsigned char *)cnoncebuf,
-                           sizeof(cnoncebuf));
+    char cnoncebuf[12];
+    result = Curl_rand_bytes(data,
+#ifdef DEBUGBUILD
+                             TRUE,
+#endif
+                             (unsigned char *)cnoncebuf,
+                             sizeof(cnoncebuf));
     if(result)
       return result;
 
     result = Curl_base64_encode(cnoncebuf, strlen(cnoncebuf),
                                 &cnonce, &cnonce_sz);

Reply:
in that case the cnoncebuf is no longer null terminated so the Curl_base64_encode input len parameter would have to be changed as well from strlen(cnoncebuf) to sizeof(cnoncebuf)

Reply:
Nice catch @jay!

Reply:
Updated diff:

~~~diff
--- a/lib/vauth/digest.c
+++ b/lib/vauth/digest.c
@@ -707,17 +707,21 @@ static CURLcode auth_create_digest_http_message(
   memset(hashbuf, 0, sizeof(hashbuf));
   if(!digest->nc)
     digest->nc = 1;
 
   if(!digest->cnonce) {
-    char cnoncebuf[33];
-    result = Curl_rand_hex(data, (unsigned char *)cnoncebuf,
-                           sizeof(cnoncebuf));
+    char cnoncebuf[12];
+    result = Curl_rand_bytes(data,
+#ifdef DEBUGBUILD
+                             TRUE,
+#endif
+                             (unsigned char *)cnoncebuf,
+                             sizeof(cnoncebuf));
     if(result)
       return result;
 
-    result = Curl_base64_encode(cnoncebuf, strlen(cnoncebuf),
+    result = Curl_base64_encode(cnoncebuf, sizeof(cnoncebuf),
                                 &cnonce, &cnonce_sz);
     if(result)
       return result;
 
     digest->cnonce = cnonce;